cvat/apps/engine/cache.py

```diff
@@ -65,63 +70,166 @@
 
 
 class MediaCache:
-    def __init__(self) -> None:
-        self._cache = caches["media"]
-
-    def _get_checksum(self, value: bytes) -> int:
+    _QUEUE_NAME = settings.CVAT_QUEUES.CHUNKS.value
+    _QUEUE_JOB_PREFIX_TASK = "chunks:prepare-item-"
+    _SLEEP_TIMEOUT = 0.2
+    _CHUNK_CREATE_TIMEOUT = 50
+    _CACHE_NAME = "media"
+    _RQ_JOB_RESULT_TTL = 60
+    _RQ_JOB_FAILURE_TTL = 3600 * 24 * 14 # 2 weeks
+    _PREVIEW_TTL = 3600 * 24 * 7
+
+    @staticmethod
+    def _cache():
+        return caches[MediaCache._CACHE_NAME]
+
+    @staticmethod
+    def _get_checksum(value: bytes) -> int:
         return zlib.crc32(value)
 
     def _get_or_set_cache_item(
-        self, key: str, create_callback: Callable[[], DataWithMime]
+        self, key: str,
+        create_callback: Callable[..., DataWithMime],
+        *args: Any,
+        cache_item_ttl: int = None,
+        **kwargs: Any,
     ) -> _CacheItem:
-        def create_item() -> _CacheItem:
-            slogger.glob.info(f"Starting to prepare chunk: key {key}")
-            item_data = create_callback()
-            slogger.glob.info(f"Ending to prepare chunk: key {key}")
+        item = self._get_cache_item(key)
+        if item:
+            return item
 
-            item_data_bytes = item_data[0].getvalue()
-            item = (item_data[0], item_data[1], self._get_checksum(item_data_bytes))
-            if item_data_bytes:
-                self._cache.set(key, item)
+        return self._create_cache_item(
+            key,
+            create_callback,
+            *args,
+            cache_item_ttl=cache_item_ttl,
+            **kwargs,
+        )
 
-            return item
+    def _get_queue(self) -> rq.Queue:
+        return django_rq.get_queue(self._QUEUE_NAME)
 
-        item = self._get_cache_item(key)
-        if not item:
-            item = create_item()
+    def _make_queue_job_id_base(self, key: str) -> str:
+        return f"{self._QUEUE_JOB_PREFIX_TASK}{key}"
+
+    @staticmethod
+    def _drop_return_value(func: Callable[..., DataWithMime], *args: Any, **kwargs: Any):
+        func(*args, **kwargs)
+
+    @classmethod
+    def _create_and_set_cache_item(
+        cls,
+        key: str,
+        create_callback: Callable[..., DataWithMime],
+        *args: Any,
+        cache_item_ttl: int = None,
+        **kwargs: Any,
+    ) -> DataWithMime:
+        item_data = create_callback(*args, **kwargs)
+        item_data_bytes = item_data[0].getvalue()
+        item = (item_data[0], item_data[1], cls._get_checksum(item_data_bytes))
```

Comment:
You likely need to add a generation timestamp into cache entries. This is needed to check for outdated entries in case of honeypot rerolls in the task. Assume here are 2 threads: server and worker

1. user requests a chunk from the server
2. worker starts preparing a missing chunk
3. user changes honeypots in a task job
4. worker finishes, the chunk stored is incorrect now
5. user requests a chunk from the server and receives an incorrect chunk